### ISSUE TITLE
Version number clarification

### ISSUE THREAD

**Issue by wolfv** [OpenImageIO/oiio#2474]
I was looking through the changelog and I am still unsure for the rationale to use what looks like 4-segment version numbers? Aren't 3 enough?

I help maintain the OpenImageIO package for conda-forge and we usually have 3 version segments + a build number (and an optional epoch) and now I am not sure what to do for OpenImageIO.

How to interpret the last number? As a doubly-minor version?

**Comment by lgritz** [comment]
Oh, the tangled web we weave.

Three was more than enough for our formal releases of supported release branches. The second number changing (e.g., 2.3 -> 2.4) indicates changes in ABI or dependency compatibility, whereas the third number (2.3.1 -> 2.3.2) may have fixes or enhancements, but only in ways where something compiled and linked against 2.3.1 will work unaltered (and without recompilation) with 2.3.2.

Master isn't a "real release." Master is named, for example, 2.4.x, if that's the next release, but it's understood that until we actually make the release branch, it is allowed to totally break compatibility at any time. At some point we branch, let's say at 2.4.3, after which point all further 2.4.x, where x >= 3, have the usual release ABI stability guarantees.

The reason we recently added a 4th number is because our studio (and others) build for production from the master branch. It can break compatibility, but yet it simply is a reality that people are relying on it. So we wanted a way to indicate to those users how they can distinguish between compatibility-breaking and -preserving changes -- within master.

So the rule is like this:  Major.minor.patch.tweak

**For supported release branches**, patch can increment with fixes or enhancements that do not break backwards compatibility, tweak is almost never used at all, but might be to indicate a hot fix that changes neither forwards nor back compatibility in any way. A non-back compatible change, or a change in dependency requirements, will increment at least the minor version and will be a totally different release branch.

**For master/development branch**, which all must be the same minor release by definition, we increment the patch when there is a compatibility break, and update the tweak when things are added that are still back-compatible.

The bottom line is that if you are only packaging supported releases, 3 numbers is all you need and you can probably get away with ignoring the 4th. 

That 4th number is a way of communicating compatible and incompatible changes to people who are working solely out of the master or development branch.

**Comment by wolfv** [comment]
Ok cool, I am just going to package releases with a `0` at the end :) Thanks for the clarification!

**Comment by wolfv** [comment]
I just updated the package on conda-forge: https://github.com/conda-forge/openimageio-feedstock

If you are interested in helping out to maintain that package, feel free to let me know :) conda is a pretty nice cross-platform package manager (but I guess you have your own solutions).

**Comment by lgritz** [comment]
Hi, @wolfv. 

I am familiar with Conda and we have experimented with it a bit at our facility.
I can definitely get in the habit of remembering to submit a PR to Conda when we do releases of OIIO, to keep it up to date.

I noticed that the Conda build recipe sets -DOIIO_BUILD_TOOLS=OFF and -DUSE_PYTHON=OFF. Is this a result of policy that no components are built aside from C++ linkable libraries? Because I think that if these two were enabled, it would (a) be very helpful to downstream dependencies who rely on either the python or command line tools for their own builds or tests, and (b) it would allow us to recommend Conda as one of the "easy ways" to build the whole shebang including dependencies.

**Comment by wolfv** [comment]
Hi Larry, that’s awesome to hear.

Regarding python: I think we should split the pure library part and the bindings. There is an open issue about this on the feedstock where i reference the GDAL feedstock: they achieve multiple outputs, one for pure lib one for python bindings.

Regarding the command line tools: yeah, let’s enable them!

Also at some point trying to do a win build would be nice...

**Comment by wolfv** [comment]
This is the issue I was referencing (yesterday I was on mobile, sorry)

https://github.com/conda-forge/openimageio-feedstock/issues/6
